### ISSUE TITLE
Feature request: search by metadata

### ISSUE THREAD

**Issue by dmitrynovik** [stripe/stripe-dotnet#1148]
It is handy that Stripe allows custom key / value pairs as metadata.

However, the metadata is not included in the search API. For example, I want to find my customer by metadata tag I attached before. That would be nice to have.

**Comment by remi-stripe** [comment]
@dmitrynovik That would definitely be a great feature to add in the API but it is currently not supported. It's not a limitation of stripe-dotnet itself here and simply that our API does not yet support something like this.

I'm going to close out the issue as it's not a bug in the library but we'd definitely like to support more powerful search in the API in the future.

**Comment by ideerge** [comment]
I want to volunteer to raise this issue again, if Stripe is supposed to be a white-label, without this feature, too much caching is required. multi metadata search is a bit of real requirement for us

**Comment by remi-stripe** [comment]
@ideerge That's definitely something we hear often and we're looking into adding support for this. No clear timeline yet but hopefully something we can update you about in the future!

**Comment by markmakes** [comment]
Agreed, would make working with Stripe immeasurably more extensible!

**Comment by spit4520** [comment]
Hey guys, its been two years since this "issue" was initially raised. I was wondering if there was any timeline on this functionality?

Thanks!

**Comment by remi-stripe** [comment]
@spit4520 We don't have any update for this functionality at the moment unfortunately. It's definitely something we're thinking of but no implementation is planned for now

**Comment by maggieWrnch** [comment]
as

**Comment by JefferyHus** [comment]
wow 2 years and the function is not yet there. So why are we sending metadata if we can't use it in a simple search? Unless you guys are saving a serialized value of metadata, it should be easy to add this.

**Comment by remi-stripe** [comment]
@JefferyHus There are a lot of usage for metadata that don't involve an API lookup. From helping analyze data in events, searching in the Dashboard or doing reconciliation in Sigma or Dashboard exports, metadata is widely used already.

We do understand that an API lookup would be really useful and it is definitely a common request that we'd like to support in the future. This would likely be implemented through a different API or pattern that would look more like a high level search and not an explicit filter on each List API endpoint. No firm plans for now though we've been looking into it more this year!

**Comment by jamieb22** [comment]
I add my vote to this feature. We are also need dire need of metadata search.

**Comment by mcfarljw** [comment]
This would also be helpful for us given how our subscriptions work. We include a reference to the user's id from out client in the metadata and it'd be great to be able to do a quick query to see if the account matches a customer before creating a number one since email addresses aren't always unique in our system.

**Comment by bahrianonline** [comment]
One more vote to add this feature. We need this feature to be able to search the customer by using our system Id instead of saving the Stripe's customer id.

**Comment by lulhum** [comment]
Same, would unlock a lot of possibilities.

**Comment by yahermann** [comment]
+1 on this feature!   

Although I think full-fledged metadata search is a big project.

A much easier-to-implement alternative is a searchable user-defined id field, i.e., `Charge.user_defined_id`, `Customer.user_defined_id`, etc., to enable integrations to fully cross-correlate Stripe rows to local database rows, and therefore enable full transaction safety.

**Comment by lucasbasquerotto** [comment]
@yahermann For single payments we charge after it's created and for subscriptions we create with a coupon with 100% discount, and charge after we inserted in our database, so as to ensure transaction safety (even if an error occurs in the middle, if it's before we inserted it, it wasn't charged yet, and if we already inserted in our database, it's safe, we just make sure that the record is not lost in a rollback), but a custom searchable id would make it much easier and straightforward, and avoid having to use these workarounds.

**Comment by yahermann** [comment]
Right, it's a chicken & egg problem. 

If you create the Stripe.Charge first, and there's a subsequent error in creating a corresponding local database row, then you end up with what we call an "orphaned" charge, meaning funds are moving in Stripe-land, but we have no record of it locally.  Not good!!!!

Conversely, if you create the local record first, followed by the Stripe.Charge, then you need some way to first verify -- before creating the Stripe.Charge -- that it has not already been created.  That's where a searchable `Charge.user_defined_uuid` becomes crucial. Currently, we query the API for **ALL** recent Charges, and then look at the metadata of each for the `user_defined_uuid` corresponding to the local database record.   We employ all sorts of trickery, including caching, to make this as quick as possible (remember, the user just clicked "Submit Payment" and is patiently waiting............. ) but it's still a pain and not completely fireproof.  And no, the idempotency layer is not an adequate workaround, because among other reasons, it expires after 24 hours.

**Comment by lucasbasquerotto** [comment]
You can solve the chicken egg problem when creating a payment intent setting both the [confirm](https://stripe.com/docs/api/payment_intents/create#create_payment_intent-confirm) parameter to `false` and the [confirmation_method](https://stripe.com/docs/api/payment_intents/create#create_payment_intent-confirmation_method) parameter to `manual`. Just do the following flow:

1) Create a record of the payment in your database with an empty field for the Stripe id.
2) Create the payment intent in Stripe with the above properties (`confirm` and `confirmation_method`) and associate your record id in a metadata field.
3) Update the record in your database with the Stripe id.
4) Confirm the payment in Stripe: https://stripe.com/docs/api/payment_intents/confirm (the payment will be charged only after this step)

The above solves the problem with failed transactions, as long as you make sure that the changes you do in the database doesn't rollback even in case of a failed transaction (because the Stripe payment will be charged only after you already associated the Stripe id with your payment record in your database). If you try to do everything in a single transaction, you could lose the record in your database (which would be bad). **The database transaction that does step 3 should end before step 4.**

For subscriptions, I follow a similar approach, but using coupons with 100% discount, and remove the coupon after I associate the subscription at Stripe with the subscription record in our database.

That said, these solutions, especially with subscriptions and coupons, are kinda ugly and hackish, but **they work** for what they are supposed to do: avoid failed transactions charge a payment in Stripe without the Stripe id being associated with a record in our database.

A searchable id **would make the solution much cleaner**, easier and straightforward, tough.

**Comment by yahermann** [comment]
Thanks for providing this workaround.

Our workaround is (grunt) rote search!!  We list all Charges, and look for our uuid within the metadata of each retrieved Charge.  The only reason this works with any semblance of "reasonableness" is because we set a Customer Id on all charges, which enables us to filter by Customer when we list all charges.  Still, this eats up quite a few precious seconds from the all-important "Submit Payment" flow.

**Comment by rnrnstar2** [comment]
I also manage user Ids with metadata.
I would like to be able to obtain user payment information all at once.
I hope you will enable the search in metadata as soon as possible.

**Comment by KumarAbhirup** [comment]
Why is this issue closed, we can all confirm we need this feature 👍

**Comment by remi-stripe** [comment]
@KumarAbhirup Github issues should be limited to bugs with the library itself. Searching via metadata is a product feature request that we recommend raising with our support team instead: https://support.stripe.com/contact

**Comment by jamieb22** [comment]
Kumer sorry this is a lame excuse. You work for Stripe yes? Then raise it. Be an active employer, looking to enhance the value of your company.

**Comment by KumarAbhirup** [comment]
Thanks @remi-stripe, contacted to product support 💯

**Comment by lucasbasquerotto** [comment]
@jamieb22 I think you are mistaken, it seems that the one more likely working for stripe is remi-stripe (after all he is a contributor and seems to be answering on behalf of Stripe). Maybe you just wrote the wrong name, though. 

@remi-stripe About contacting stripe to ask for this feature, I think that here on Github is a better place to ask, because several people can track the issue, while having to contact stripe each person would have to contact separately, without knowing if other people have the same issue, and without having a way to track the issue (to know if and when it is implemented).

It doesn't need to be here on Github, though (because, as you said, Github issues should be limited to bugs). It could be a discourse forum where you can watch a topic or any such platform that allow users to make requests, see already opened requests (to avoid duplication and help to know if a feature is very requested), and mainly to allow track and watch the request in a transparent and practical way (so as to be notified when the feature is implemented, or that the feature will not be implemented).

That said, if there will be not feedback here (and no action will be taken based on what is asked here) then it is no better to ask here than to ask in another place (and contacting support directly may end up being the only way).

**Comment by clement911** [comment]
Totally agree with @lucasbasquerotto , Stripe could really do with some sort of public community forum.
Something like Shopify has: https://community.shopify.com
I'm sure they've thought of it so there must be a reason why they haven't done it yet.

**Comment by mcfarljw** [comment]
I have reached out to Stripe support referencing this issue and got the following reply:
```
I will go ahead and pass this over to our team, was there anything else?
```

To the point of @lucasbasquerotto using Github to track things is more transparent and easier to let the masses know the status of things. Based on the reply I got from support I just feel like my request went into the void for something that has clearer garnered quite a bit of support on Github as a desired feature.

@remi-stripe Things might have changed since 2 years ago when this was initially opened but looking at the current open issues marked with the label `future` it seems like other feature requests with far less user support are being entertained and not being closed with a response to contact support. https://github.com/stripe/stripe-dotnet/issues?q=is%3Aissue+Feature+is%3Aopen+label%3Afuture

**Comment by remi-stripe** [comment]
@mcfarljw The issues tagged as `future` are about improving the core of the library itself. This request is a product feature itself which is separate from the library. If we ship support for listing via metadata it would immediately be available in all our client libraries.

I do hear you on the tracking, but I disagree as this is not really scalable. Github issues are limited to bug or feature requests about the specific library itself.

We are not ignoring those requests and we track them but this one is not something we plan to ship in the near future. It is asked occasionally but a lot less than many other core feature requests. We are working on an alternative approach though, something akin to a search API across all resources and hope that we can release this next year which could address the metadata problem but no firm timeline for now.

**Comment by mcfarljw** [comment]
@remi-stripe fair enough. I can get behind it not being something specific to the library but rather a higher-level feature that needs to be created so the library can support it.

I guess the reason why I thought it's something that just hasn't been added to the library and already is supported is that if I go to https://dashboard.stripe.com/dashboard and use the search feature I can search for things by metadata. For example, we have a `user` metadata property and when I search using one it returns the customer associated with that metadata. I realize this might be a feature specific to the website dashboard that hasn't made it to the public API but it does seem to be something that can be done and is being used already.

**Comment by justinputney** [comment]
I'm looking for this feature too (with the PHP API).

**Comment by dobrud** [comment]
I would love to see this feature as well (with the stripe-node)

**Comment by Salman2301** [comment]
Yes please! Thank you

**Comment by agostinofiscale** [comment]
I think that if this feature will be available it would super useful.

**Comment by dkovacevic** [comment]
I believe this is what you are all looking for: (Java code)
Stripe API 2020-08-27
This code will find Customer object that has `metadata` key `teamId` equals {teamId}
```
public Customer search(TeamId teamId) throws IbisException {
            final String url = String.format("%s%s", Stripe.getApiBase(), "/v1/search");
            final Map<String, Object> params = new HashMap<>();
            params.put("query", String.format("metadata:teamId=%s is:customer", teamId));
            final CustomerCollection collection = ApiResource.request(ApiResource.RequestMethod.GET,
                    url,
                    params,
                    CustomerCollection.class,
                    RequestOptions.getDefault());
            return collection.getData().stream()
                    .findFirst()
                    .orElseThrow(() -> new MissingCustomerException(teamId));
    }
   ```

**Comment by yoroshiku-zz** [comment]
We're waiting for this feature too (PHP, Python).

**Comment by lucasconstantino** [comment]
Some rough implementation using Node API + TypeScript:

```ts
import Stripe from 'stripe'

// Extended 
type ExtendedStripe = Stripe & {
  search: (config?: { query: string }) => Stripe.ApiListPromise<Stripe.Customer>
}

/**
 * Inject a "search" method to consume from /v1/search.
 */
const injectSearch = (stripe: Stripe) => {
  const SearchResource = Stripe.StripeResource.extend({
    path: 'search',

    search: Stripe.StripeResource.method({
      method: 'GET',
      methodType: 'list',
      path: '',
    }),
  })

  // @ts-ignore create search resource instance
  const searchResource = new SearchResource(stripe) as ExtendedStripe

  // @ts-ignore inject root search method
  stripe.search = searchResource.search.bind(searchResource) as ExtendedStripe['search']

  return stripe as ExtendedStripe
}

const stripe = injectSearch(
  new Stripe(process.env.STRIPE_API_KEY, {
    apiVersion: '2020-08-27',
  })
)

export { stripe }
```

Usage:

```ts
const { data } = await stripe.search({ query: 'metadata:key=value' })

```

**Comment by remi-stripe** [comment]
@lucasbasquerotto unless you're in the "secret" beta, this API isn't really supported and doesn't exist and we discourage anyone from trying to use this. If you're in the "secret" beta, it's mostly secret :p

**Comment by lucasbasquerotto** [comment]
@remi-stripe I think you got the wrong guy here ( ͡° ʖ̯ ͡°) 

You probably meant @lucasconstantino

**Comment by remi-stripe** [comment]
Ah yes, sorry, one day I'll understand how auto-completion on github works 😓

**Comment by derekl-stripe** [comment]
@lucasconstantino, please send us an e-mail at support-search-api@stripe.com if you're willing to try out the new API as a beta user. We initially started small on the beta initially but can start gathering more feedback from a larger cohort of users.

**Comment by pfcodes** [comment]
Need this!

**Comment by derekl-stripe** [comment]
@pfcodes, please send us an e-mail at support-search-api@stripe.com if you're willing to use the new API as a beta user.

**Comment by vivekayer** [comment]
The documentation says you can search by metadata but I'm not able to.  Any idea why?

[https://stripe.com/docs/dashboard/search#metadata-searches](https://stripe.com/docs/dashboard/search#metadata-searches)

**Comment by remi-stripe** [comment]
@vivekayer That doc is about the Dashboard, and metadata search does work there. I'd recommend talking to our support team for help with the Dashboard: https://support.stripe.com/contact/email

**Comment by hn1** [comment]
Forget the dashboard. We are programers. Definitely a required feature.

**Comment by Shaunwild97** [comment]
+1 This is required for me too.

**Comment by Tanner-Scadden** [comment]
+1

**Comment by nerique** [comment]
+1

**Comment by nmtitov** [comment]
Any updates? This feature is crucial to establish connection between Stripe client ids and our internal CRM user ids

**Comment by remi-stripe** [comment]
@nmtitov No update for now no. We are working on a Search API and have a beta ongoing, but it's not ready to ship publicly yet.

**Comment by elDudarino** [comment]
I VOTE for this! Having to grab all customers and check each meta individually is not great.

**Comment by drmrbrewer** [comment]
+1 for filtering based on metadata... specifically for listing/filtering customer: https://stripe.com/docs/api/customers/list

**Comment by derekl-stripe** [comment]
I'm happy to announce that we are now in public beta for the Search API where you can search by metadata and several other fields. We are also releasing beta SDKs for Node.js and Java that can access the Search API.

You can take a look at the docs at https://stripe.com/docs/search-api and drop us an email at search-api@stripe.com with the ID(s) of the account you want to enable when you're ready to use the API.

**Comment by open-nikita** [comment]
@derekl-stripe thanks! our beta testing experience using search API was nothing but a pleasure. Looking forward to switching to node sdk!

**Comment by amytsai-stripe** [comment]
Thanks to everyone who commented in this thread, and expressed interest in the beta. The Stripe Search API is now available for use, and has SDK support! Read the docs at https://stripe.com/docs/search.
